Implement DDP session resumption for graceful reconnects

vlasky opened the pull request:
## Summary

This PR implements DDP session resumption, allowing clients to automatically resume their previous session after a temporary network disconnect without re-triggering `onConnection` callbacks or losing their connection ID.

**This PR supersedes PR #13378** with a completely rewritten implementation that includes comprehensive tests to give everyone confidence in its correctness.

## Key Features

- **Automatic session resumption**: When a client reconnects within the grace period (default 15 seconds), their session is seamlessly restored
- **Configurable grace period**: `Meteor.server.options.disconnectGracePeriod` (default: 15000ms)
- **Message queue limit**: `Meteor.server.options.maxMessageQueueLength` (default: 100) prevents memory leaks during disconnects
- **Graceful vs ungraceful disconnect handling**: Only ungraceful disconnects (network loss, browser close) allow resumption; graceful disconnects (explicit logout, server kick) do not
- **Message count verification**: Server validates client's received message count to ensure no messages were lost

## Implementation Details

### Client-side changes (ddp-client):
- Track `receivedCount` of messages from server (excluding ping/pong)
- Send `receivedCount` in reconnect messages
- Handle `session_resumed` vs new session responses
- Send `disconnect` message before intentional disconnects

### Server-side changes (ddp-server):
- Maintain `sentCount` of messages sent to each client
- On disconnect, queue session for removal after grace period
- On reconnect with matching session ID and count, resume session
- Prevent resumption for graceful disconnects and server-initiated closes
- Guard against double cleanup and memory leaks

## Comprehensive Test Coverage

### Server-side tests (7 new tests):
- `DDP resumption: unexpected disconnect preserves session` - verifies grace period works
- `DDP resumption: graceful disconnect removes session immediately` - verifies client disconnect msg works
- `DDP resumption: server-initiated close removes session immediately` - verifies server kicks work
- `DDP resumption: onConnection not called on resume` - verifies hook behavior
- `DDP resumption: server close prevents resumption` - verifies server kicks are not resumable
- `DDP resumption: graceful disconnect prevents resumption` - verifies client disconnects are not resumable
- `DDP resumption: count mismatch creates new session` - verifies mismatched counts create new session

### Client-side tests (5 new tests):
- `receivedCount tracking` - verifies count increments and ignores ping/pong
- `receivedCount sent on reconnect` - verifies reconnect includes count
- `receivedCount reset on new session` - verifies reset on new session
- `receivedCount preserved on session resume` - verifies count continues on resume
- `disconnect sends disconnect message` - verifies disconnect() sends msg

## Documentation

Updated v3-docs to document the new feature and configuration options.

## Test plan

- [x] All existing ddp-server tests pass
- [x] All existing ddp-client tests pass
- [x] New DDP resumption tests pass
- [x] Verified with automated Puppeteer testing (page reload simulation)

Closes #13378

vlasky committed:
Implement DDP session resumption for graceful reconnects

This feature allows Meteor clients to resume their DDP connections after
temporary disconnections without re-establishing full session state.

Key changes:
- Server tracks sent message count per session (sentCount)
- Client tracks received message count (receivedCount)
- On reconnect, client sends receivedCount with session ID
- Server validates counts match to allow session resumption
- New disconnect message allows graceful disconnects
- Grace period (default 15s) keeps session alive after disconnect
- Message queue (max 100) buffers messages during disconnect

Server options:
- Meteor.server.options.disconnectGracePeriod (default: 15000ms)
- Meteor.server.options.maxMessageQueueLength (default: 100)

This significantly reduces server CPU spikes when clients reconnect
(e.g., after Google Cloud Run timeouts) by avoiding full session
recreation and data re-fetch.

Based on PR #13378. Rebased onto current devel branch with refactored
ddp-client handlers.

Co-authored-by: Jan Dvorak <AsciiDoctor@outlook.com>
Co-authored-by: Valentin Slatineanu <valentin.slatineanu@gmail.com>
Co-authored-by: zodern <zodern@icloud.com>

vlasky committed:
Fix flaky async publish cursor test on page reload

The test was failing when the browser page was reloaded during test runs
because the subscription ready callback could fire before the added
messages arrived. This can happen when a previous test run was
interrupted and the server is still processing the old session.

The fix adds a polling mechanism to wait for data to arrive before
asserting, with a 5-second timeout.

italojs commented on packages/ddp-server/livedata_server.js:
```diff
@@ -391,6 +424,13 @@ Object.assign(Session.prototype, {
       return;
     }
 
+    if (msg_in.msg === 'disconnect') {
```
why not `msg_in.msg === 'disconnect' && (msg_in.msg in self.protocol_handlers) ?

italojs replied:
if we validate it in the entrance, we can avoid this condition

vlasky replied:
Good catch. The inner check is indeed redundant since `disconnect` is always defined in `protocol_handlers`. I'll simplify this to call the handler directly:

```javascript
if (msg_in.msg === 'disconnect') {
  return self.protocol_handlers.disconnect.call(self, msg_in, () => {});
}

This also addresses your later comment about validating at the entrance.

italojs commented on packages/ddp-client/common/livedata_connection.js:
```diff
@@ -102,6 +106,7 @@ export class Connection {
 
     self._heartbeatInterval = options.heartbeatInterval;
     self._heartbeatTimeout = options.heartbeatTimeout;
+    self._ignoredMsgsForSessionOutOfDateCheck = ['ping', 'pong'];
```
why events like 'connected', 'nosub', matter here?

vlasky replied:
Yes, `connected` and `nosub` are intentionally counted in `receivedCount`. Only `ping/pong` are excluded because they're heartbeat noise.

Counting `connected` and `nosub` is important because:
- `connected`is sent once per connection and matters for session state
- `nosub` indicates subscription failure - losing this could cause client/server desync

If we excluded them, the client and server counts could diverge, forcing unnecessary session resets on reconnection.

italojs commented on packages/ddp-server/livedata_server.js:
```diff
@@ -1424,13 +1480,63 @@ Object.assign(Server.prototype, {
     // Yay, version matches! Create a new session.
```
// Yay, version matches! Create/Renew a session.

vlasky replied:
Good suggestion! I'll update the comment to reflect that we may either resume an existing session or create a new one.

italojs commented on v3-docs/docs/api/meteor.md:
```diff
@@ -910,16 +910,17 @@ contains the following fields:
-> connection `id`.
+> In previous versions of Meteor, when a client reconnects to the server (such as after temporarily losing its Internet connection), it will get a new connection each time. The `onConnection` callbacks will be called again, and the new connection will have a new connection `id`.
+
+> With the new client reconnection feature ([DDP resumption](https://github.com/meteor/meteor/pull/13378)) introduced in Meteor version 3.4, the client will attempt to automatically resume the previous connection to the server without calling the `onConnection` callback again and the connection will still keep the previous connection `id`. This functionality is controlled by the following new ser ...
```
IMO, we shouldn't specify about "previous version" or "new client feature" since it can become deprecated one day. I recommend using something like  " >=3.5 introduces..." 

another thing I would like to recommend is to add a subtitle here `## reconnection`

vlasky replied:
Agreed on both points. I'll:

1. Rephrase to avoid time-sensitive language
2. Add a ## Reconnection subsection for better organization

I'll also update the PR link to reference this PR instead of #13378.

italojs commented on packages/ddp-server/livedata_server.js:
```diff
@@ -269,77 +278,101 @@ Object.assign(Session.prototype, {
+      self.inQueue = null;
+      self.collectionViews = new Map();
+
+      if (self.heartbeat) {
```
this code is duplicated, we can move for a function

vlasky replied:
Agreed. I'll extract it to a helper method like `_stopHeartbeat()`.

italojs commented on packages/ddp-server/livedata_server.js:
```diff
@@ -391,6 +424,13 @@ Object.assign(Session.prototype, {
+      if (msg_in.msg in self.protocol_handlers) {
+        // we want to pre-empt the queue - a disconnect is imminent.
+        return self.protocol_handlers[msg_in.msg].call(self, msg_in, () => {});
+      }
```
I guess the question here is: should we accept invalid msg inside `processMessage` ?

vlasky replied:
The current design already handles invalid messages appropriately. If a message type isn't in `protocol_handlers`, line 475 sends `sendError('Bad request', msg)` back to the client.

For `disconnect`, we pre-empt the queue (similar to ping/pong) because it's a control message signaling imminent connection closure. With the simplified code from the earlier comment, we call the handler directly without the redundant existence check, so invalid messages never reach this path for `disconnect`.

italojs commented on packages/ddp-server/livedata_server.js:
```diff
@@ -1424,13 +1480,63 @@ Object.assign(Server.prototype, {
+    // was disconnected and hasn't been reconnected to yet.
+    if (existingSession && existingSession.sentCount === msg.receivedCount && existingSession._removeTimeoutHandle) {
+      Meteor.clearTimeout(existingSession._removeTimeoutHandle);
+      delete existingSession._removeTimeoutHandle;
```
please set to undefined, v8 do not activate JIT for this piece of code when we use delete statement since it changes the object schema at runtime


benchmark: https://akashsingh.blog/javascript-delete-operator-might-cause-some-unexpected-performance-issues?source=more_articles_bottom_blogs (tested locally into node 25, the results is true)

ref: `if many properties get added and deleted from an object, it can generate a lot of time and memory overhead to maintain the descriptor array and HiddenClasses` https://v8.dev/blog/fast-properties 

ref2: https://dev.to/maxprilutskiy/hidden-classes-the-javascript-performance-secret-that-changed-everything-3p6c

vlasky replied:
Yes you are right. I will make this change.

vlasky commented on v3-docs/docs/api/meteor.md:
```diff
@@ -910,16 +910,17 @@ contains the following fields:
+
+### Meteor.server.options.maxMessageQueueLength
+
+Determines how many messages we should queue during a non-graceful disconnect before we destroy the session, to insure against memory leaks. Defaults to `100`.
```
I will rewrite this.

vlasky commented on packages/ddp-server/livedata_server.js:
```diff
@@ -269,77 +278,101 @@ Object.assign(Session.prototype, {
+        self.heartbeat.stop();
+        self.heartbeat = null;
+      }
+
```
While it appears redundant, I'm keeping the defensive check in the callback. The callback executes asynchronously after `_removeSession` and the session state could theoretically change in edge cases. The check is cheap and provides defense-in-depth. However, per italojs's earlier comment, I will extract this to a `_stopHeartbeat()` helper to reduce duplication while keeping both call sites.

vlasky commented on packages/ddp-server/livedata_server_tests.js:
```diff
@@ -593,4 +628,344 @@ function getTestConnections(test) {
+  "livedata server - DDP resumption: server-initiated close removes session immediately",
+  async function (test) {
+    await withTestGracePeriod(async () => {
+      const { clientConn, serverConn } = await getTestConnections(test);
```
Will fix.

vlasky commented on packages/ddp-client/common/livedata_connection.js:
```diff
@@ -1081,11 +1086,12 @@ export class Connection {
 
   close() {
-    return this._stream.disconnect({ _permanent: true });
+    return this.disconnect({ _permanent: true });
```
The `_permanent` flag is used by the underlying stream (socket-stream-client), not the DDP server. It controls whether the stream should attempt to reconnect after closing. Keeping it ensures `close()` is truly permanent while `disconnect()` allows for potential reconnection. I'll add a comment to clarify this.

vlasky commented on packages/ddp-server/livedata_server.js:
```diff
@@ -1252,6 +1295,19 @@ Server = function (options = {}) {
     respondToPings: true,
     defaultPublicationStrategy: publicationStrategies.SERVER_MERGE,
+    /**
+     * @summary How many messages should we queue during a non-graceful disconnect before we kill the session (to insure against memory leaks).
```
I will rewrite this.

vlasky committed:
Address PR review feedback

- Fix wording: "insure" → "to help prevent" in docs and JSDoc
- Extract heartbeat cleanup to _stopHeartbeat() helper method
- Simplify disconnect handling: remove redundant condition check
- Update documentation: add Reconnection section, use "Meteor 3.5+"
  instead of time-sensitive language, update PR link to #14051
- Update comment at line 1480 to reflect session resume/create behavior
- Replace delete statements with = undefined for V8 JIT optimization
- Remove unused variable clientConn in test
- Add comment clarifying _permanent flag purpose in close()

italojs committed:
Merge branch 'devel' into feature/ddp-resumptions-rebased

italojs commented:
Just saw your branch is updated with 3.4.1, I'll maintain it pointed to 3.4.1 just for review propose, but when do you have a time, can you remove the `[d09220f](https://github.com/meteor/meteor/pull/14051/commits/d09220f579f2be7df1967abfd605b85c1807fc1b)` commit, please?

``` 
git pull
git reset --hard HEAD~1
git push -f
```

vlasky commented:
> Just saw your branch is updated with 3.4.1, I'll maintain it pointed to 3.4.1 just for review propose, but when do you have a time, can you remove the `[d09220f](https://github.com/meteor/meteor/pull/14051/commits/d09220f579f2be7df1967abfd605b85c1807fc1b)` commit, please?
> 
> ```
> git pull
> git reset --hard HEAD~1
> git push -f
> ```

done

italojs committed:
Merge branch 'release-3.5' into feature/ddp-resumptions-rebased

italojs committed:
Merge branch 'release-3.5' into feature/ddp-resumptions-rebased
